tracer/src/Datadog.Trace/Tracer.cs

```diff
@@ -213,6 +213,11 @@ public static Tracer Instance
             {
                 lock (_globalInstanceLock)
                 {
+                    if (_instance is ILockedTracer)
+                    {
+                        throw new Exception("The current tracer instance cannot be replaced.");
+                    }
+
```

Comment:
Under what circumstances could this happen? 🤔 Do we definitely want to throw (rather than log an error and ignore the set), as that would bring down the test process?

Reply:
I'm not sure, the singleton setter is part of the public API, If the process start with CI Visibility mode we have to prevent the tracer change due to billing implications. It's true we can ignore the set instead of throwing, we have both options. My first approach was this, just failing any test that tries to replace the `Tracer.Instance` as an unsupported scenario (I really don't expect to hit this scenario very often), the problem with ignoring the set is to gave false positives to customers tests without any explanation.

Reply:
I guess the question is:

* Are we primarily worried about _customers_ replacing the tracer, or _us_ in our tests?
* If it's customers we're worried about, do we want the experience that adding our tracer breaks their tests?

If we silently (with logging) ignored it, instead of throwing, then any spans they created would still be captured as part of the CI visibility spans right? Unless they had tests asserting on what the Tracer was doing, then it shouldn't change their tests? (And if they _are_ doing that, that seems bad practice, and fine to fail anyway)? 

> the problem with ignoring the set is to gave false positives to customers tests without any explanation

I don't quite understand this, what are the "false positives" you're referring to here?

Reply:
Yeah, It's about the customer replacing the tracer. 

What I'm worried about ignoring silently is changing the behaviour of a test suite without telling the user. Yes, we can log what's happening, but we don't really know if is going to fail or succeed, or maybe introduces flakiness. And the experience for a customer to retrieve the native logs in a CI is not straightforward.

In the other hand, I'm not so worried about this scenario because I don't expect customers trying to replace the tracer in a test suite.

At the end, I think I'm fine with both solutions... :shrug:

Reply:
>And the experience for a customer to retrieve the native logs in a CI is not straightforward.

That's a very good point 

>In the other hand, I'm not so worried about this scenario because I don't expect customers trying to replace the tracer in a test suite.
> At the end, I think I'm fine with both solutions... 🤷

Yeah, likewise, seems like an edge case 👍